### PR TITLE
router: add responseHeaderTransforms API

### LINKED ISSUE #15012: [router]: expose header modifcations protobufs to filters.
*Title*: Expose header modifications to filters.

*Description*:
We have the requirement to let the filter peek into what header modifications would be done to the request/response, right now the modifications are all hidden in router which is at the end of the chain.

### PR DESCRIPTION
**Pull request by ahedberg**
Signed-off-by: Ashley Hedberg <ahedberg@google.com>

Commit Message: Adds `ResponseEntry::responseHeaderTransforms` and `HeaderParser::getHeaderTransforms`.
Additional Description: These API allow filters to obtain configured response header transformations on the request path. This is useful if the response header transformations need to be sent to another (non-Envoy) component.
Risk Level: low
Testing: new and existing unit tests
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Fixes #15012

### CODE REVIEW

File: include/envoy/router/router.h
Comment (dio) [inline include/envoy/router/router.h:66]: Sorry, mind to add `@param stream_info ...` here?

File: include/envoy/http/header_map.h
Comment (dio) [inline include/envoy/http/header_map.h:76]: Probably "bar" is not the right name for this?

File: source/common/router/config_impl.cc
Comment (dio) [inline source/common/router/config_impl.cc:625]: Could you expand this comment with examples? Thanks!

File: source/common/router/header_parser.h
Comment (dio) [inline source/common/router/header_parser.h:57]: Please add "@param" here.

File: test/common/router/config_impl_test.cc
Comment (dio) [inline test/common/router/config_impl_test.cc:46]: Not sure about the movement here. I don't disagree, but should we set the same style for all?

**Review by dio — CHANGES_REQUESTED** [review]
Thank you for working on this! Looks good. Sorry for some nitpicking. Do you think we need to expose this to some docs?

The clang-tidy error seems valid.

File: include/envoy/http/header_map.h
Comment (ahedberg) [reply include/envoy/http/header_map.h]: Renamed.

File: include/envoy/router/router.h
Comment (ahedberg) [reply include/envoy/router/router.h:67]: Added.

File: source/common/router/config_impl.cc
Comment (ahedberg) [reply source/common/router/config_impl.cc]: I copied this from `RouteEntryImplBase::finalizeResponseHeaders` so don't have any good examples, but I expanded the comment a bit.

File: source/common/router/header_parser.h
Comment (ahedberg) [reply source/common/router/header_parser.h:58]: Added

File: test/common/router/config_impl_test.cc
Comment (ahedberg) [reply test/common/router/config_impl_test.cc:46]: This change follows https://abseil.io/tips/119. My experience has been that folks prefer making style cleanups when modifying nearby code than cleanup-only PRs, which is why I haven't applied this everywhere.

**Comment by ahedberg** [comment]
> Thank you for working on this! Looks good. Sorry for some nitpicking. Do you think we need to expose this to some docs?
> 
> The clang-tidy error seems valid.

Where would be a good place to add docs for this?

**Comment by ahedberg** [comment]
> Thank you for working on this! Looks good. Sorry for some nitpicking. Do you think we need to expose this to some docs?
> 
> The clang-tidy error seems valid.

ClangTidy is fixed in newest CI runs. Coverage run failed because //test/extensions/quic_listeners/quiche/integration:quic_http_integration_test failed, which I don't think is related.

**Comment by dio** [comment]
re: Docs, I think we can add a short one here: https://github.com/envoyproxy/envoy/tree/main/source/docs the content could be similar to https://github.com/envoyproxy/envoy/blob/main/source/docs/fancy_logger.md probably?

**Review by dio — APPROVED** [review]
Thanks! Pinging @envoyproxy/senior-maintainers for the second pass.

**Comment by ahedberg** [comment]
Docs added.

**Review by dio — DISMISSED** [review]
Thank you!

**Comment by mattklein123** [comment]
Can you merge main?

/wait

**Comment by ahedberg** [comment]
> Can you merge main?
> 
> /wait

Merged.

**Comment by ahedberg** [comment]
Anything else left here before merging?

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
